### PR TITLE
Exporting functionality added

### PR DESCRIPTION
**Pull request by adrosenbaum**
User can now export synthetic datasets from mutacc by making a query to
the db, and providing 'background' bam files. The synthetic dataset will
be returned as bam files enriched with variants from the query.

### CODE REVIEW

File: mutacc/builds/build_dataset.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    #Find child
+    for sample in samples:
+
+        if sample["mother"] in sample_ids \
+        and sample["father"] in sample_ids:
+
+            child = sample
+            break
+    try:
+        child
```
Comment (henrikstranneheim) [inline mutacc/builds/build_dataset.py]: What about siblings? How would these be handled? @moonso Do you have any functions in genmod or other packages that already perform these identity checks?

File: mutacc/builds/build_variant.py [hunk 2]
```diff
@@ -34,28 +34,28 @@ def find_region(self, padding):
...
                 region = {"start": self.entry.start - padding,
                           "end": self.entry.end + padding}
 
             #Find the region for variant with an 'SVTYPE' ID in the INFO field of the vcf entry
             else:
-                
+
                 #Depending on the direction of the break end, the region is extended either upstreams
                 #or downstream of the position of the break end.
                 if "[" in self.entry.ALT[0]:
```
Comment (henrikstranneheim) [inline mutacc/builds/build_variant.py]: Are the if and else condition defining the same region? Maybe I am missing something?

File: mutacc/mutaccDB/query.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        variants = mutacc_adapter.find_variants(
+                {"_id": {"$in": case["variants"]}}
+            )
+        for variant in variants:
+
+            region =  {"chrom": variant["chrom"],
+                       "start": variant["reads_region"]["start"],
+                       "end": variant["reads_region"]["end"]}
+
+            if overlapping_region(region, final_regions):
```
Comment (henrikstranneheim) [inline mutacc/mutaccDB/query.py]: Nice!

File: mutacc/utils/bam_handler.py [hunk 4]
```diff
@@ -130,30 +153,36 @@ def out_file(self):
 
         return str(self.out_name)
 
-    #This method may be implemented using picard FilterSamReads or another faster command line
```
Comment (henrikstranneheim) [inline mutacc/utils/bam_handler.py]: Have you tried bedtools or samtools?
1. bedtools intersect -abam file.bam -b filter.bed -v > filtered.bam
2. samtools view -U -L filter.bed > filtered.bam

File: mutacc/builds/build_variant.py (hunk 2 above)
Comment (henrikstranneheim) [reply mutacc/builds/build_variant.py]: Also, this looks like a very good place to use a technique called a dispatch table. Look it up and see if you can refactor the function.

File: mutacc/builds/build_dataset.py (hunk 1 above)
Comment (adrosenbaum) [reply mutacc/builds/build_dataset.py]: at the moment this only work for father-mother-child trios. I'm trying to come up with a way that would work for all cases. I thought we could discuss this in our next meeting.

File: mutacc/utils/bam_handler.py (hunk 4 above)
Comment (adrosenbaum) [reply mutacc/utils/bam_handler.py]: That might be a better option. The only problem is that we might want to filter out all the mates to the reads overlapping the regions as well.

File: mutacc/builds/build_dataset.py (hunk 1 above)
Comment (moonso) [reply mutacc/builds/build_dataset.py]: Yes I agree. This should be handled already when parsing the ped file. In that way you can store something like `sample.is_child=True`. You can look at github.com/moonso/ped_parser if you like.

File: mutacc/builds/build_dataset.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+    #Find parents
+    for sample in samples:
+
+        if sample["sample_id"] == child["father"]:
+            father = sample
+        elif sample["sample_id"] == child["mother"]:
+            mother = sample
+
+    try:
```
Comment (moonso) [inline mutacc/builds/build_dataset.py]: Same as above

**Comment by adrosenbaum** [comment]
Some changes in PR based on discussion on export functionality we had on Friday.
Now, the read names overlapping the region are found with pysam,
and excluded from the input 'background' fastq files with command
line tool seqkit, which I think works well, and fast, for the purpose.

The pedigree structure of the cases is parsed with moonso's ped_parser
using the Individual, and Family classes.

When using mutacc grep now, the user must provide option -m, stating
the family member to make a synthetic fastq file for. if -m 'affected',
A Synthetic fastq based on any affected sample from all of the cases,
regardless of single or trio, will be made.

I've made pretty large changes since the last commit, and I am sorry for
this very bulky PR which I understand is a mess to review. If you are interested
in discussing the changes let me know.

**Pull request merged**